android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java

```diff
@@ -635,4 +700,142 @@ private Tab getTab() {
+            }
+            indicatorLayout.addView(dotTextView);
+        }
+    }
+
+    // NTP related methods
+    private NTPWidgetAdapter.NTPWidgetListener ntpWidgetListener =
+            new NTPWidgetAdapter.NTPWidgetListener() {
+                @Override
+                public void onMenuEdit() {
+                    cancelTimer();
+                    openWidgetStack();
+                }
+
+                @Override
+                public void onMenuRemove(int position) {
+                    showWidgets();
+                }
+
+                @Override
+                public void onMenuLearnMore() {
+                    TabUtils.openUrlInSameTab(BRAVE_BINANCE);
+                }
+
+                @Override
+                public void onMenuRefreshData() {
+                    mBinanceNativeWorker.getAccountBalances();
+                }
+
+                @Override
+                public void onMenuDisconnect() {
+                    mBinanceNativeWorker.revokeToken();
+                }
+            };
+
+    private BinanceObserver mBinanaceObserver = new BinanceObserver() {
+        @Override
+        public void OnGetAccessToken(boolean isSuccess) {
+            Log.e("NTP", "OnGetAccessToken : " + isSuccess);
+            BinanceWidgetManager.getInstance().setUserAuthenticationForBinance(isSuccess);
+            if (isSuccess) {
+                mBinanceNativeWorker.getAccountBalances();
+            }
+        };
+
+        @Override
+        public void OnGetAccountBalances(String jsonBalances, boolean isSuccess) {
+            if (!isSuccess) {
+                BinanceWidgetManager.getInstance().setUserAuthenticationForBinance(isSuccess);
+            }
+            try {
+                BinanceWidgetManager.binanceAccountBalance =
+                        new BinanceAccountBalance(jsonBalances);
+            } catch (JSONException e) {
+                Log.e("NTP", e.getMessage());
+            }
+            // Reset binance widget to connect page
+            showWidgets();
+        };
+
+        @Override
+        public void OnGetConvertQuote(
+                String quoteId, String quotePrice, String totalFee, String totalAmount){
+
+        };
+
+        @Override
+        public void OnGetCoinNetworks(String jsonNetworks){};
+
+        @Override
+        public void OnGetDepositInfo(String depositAddress, String depositTag, boolean isSuccess){};
+
+        @Override
+        public void OnConfirmConvert(boolean isSuccess, String message){};
+
+        @Override
+        public void OnGetConvertAssets(String jsonAssets){};
+
+        @Override
+        public void OnRevokeToken(boolean isSuccess) {
+            BinanceWidgetManager.getInstance().setUserAuthenticationForBinance(!isSuccess);
+            ntpWidgetAdapter.notifyDataSetChanged();
+        };
+    };
+
+    // start timer function
+    public void startTimer() {
+        countDownTimer = new CountDownTimer(30000, 1000) {
```

Comment:
do we need to check balance only once after 30s or every 30s?

Reply:
@SergeyZhukovsky after every 30 seconds. and that's why i have cretaed functions to start and cancel the timer to control the behaviour.